google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/model.scala

```diff
@@ -77,66 +77,168 @@ object PubSubConfig {
+}
+
 /**
- *
- * @param data the base64 encoded data
- * @param messageId the message id given by server. It must not be populated when publishing.
- * @param attributes optional extra attributes for this message.
- * @param publishTime the time the message was published. It must not be populated when publishing.
+ * 'data' of [[ReceivedMessage]].
+ * @param data the base64 encoded data, if not present, attributes have to contain at least one entry
+ * @param attributes attributes for this message, if not present, data can't be empty
+ * @param messageId the message id given by server.
+ * @param publishTime the time the message was published.
  */
-final case class PubSubMessage(data: String,
-                               //Should be Option[String]. '""' is used as default when creating messages for publishing.
-                               messageId: String,
-                               attributes: Option[immutable.Map[String, String]] = None,
-                               publishTime: Option[Instant] = None) {
+final class PubSubMessage private (val data: Option[String],
+                                   val attributes: Option[immutable.Map[String, String]],
+                                   val messageId: String,
+                                   val publishTime: Instant) {
 
   def withAttributes(attributes: java.util.Map[String, String]): PubSubMessage =
-    copy(attributes = Some(attributes.asScala.toMap))
+    new PubSubMessage(data, Some(attributes.asScala.toMap), messageId, publishTime)
 
-  def withPublishTime(publishTime: Instant): PubSubMessage =
-    copy(publishTime = Some(publishTime))
+  def withData(data: String): PubSubMessage =
+    new PubSubMessage(Some(data), attributes, messageId, publishTime)
 
+  override def equals(other: Any): Boolean = other match {
+    case that: PubSubMessage =>
+      data == that.data && attributes == that.attributes && messageId == that.messageId && publishTime == that.publishTime
+    case _ => false
+  }
+
+  override def hashCode: Int = java.util.Objects.hash(data, attributes, messageId, publishTime)
+
+  override def toString: String =
+    "PubSubMessage(data=" + data + ",attributes=" + attributes + ",messageId=" + messageId + ",publishTime=" + publishTime + ")"
 }
 
 object PubSubMessage {
 
-  def apply(data: String): PubSubMessage = PubSubMessage(data, "")
+  def apply(data: Option[String] = None,
+            attributes: Option[immutable.Map[String, String]] = None,
+            messageId: String,
+            publishTime: Instant) = new PubSubMessage(data, attributes, messageId, publishTime)
 
-  def apply(data: String, attributes: immutable.Map[String, String]): PubSubMessage =
-    PubSubMessage(data, "", Some(attributes), None)
-
-  /**
-   * Java API: create [[PubSubMessage]]
-   */
-  def create(data: String) =
-    PubSubMessage(data)
+  def create(data: Option[String],
+             attributes: Option[immutable.Map[String, String]],
```

Comment:
Replace with `java.util.Map`. Also here and above replace `Option` with `Optional`.